### ISSUE TITLE
dune exec needs to add .exe on Windows

### ISSUE THREAD

**Issue by dra27** [ocaml/dune#3322]
`dune exec -- foo` on Windows needs to consider `dune exec -- foo.exe` when working out if there is a program to (re)compile. At the moment, the behaviour is mixed:

- if `_build/install/default/bin/foo.exe` has previously been built, then you get a potentially stale execution of `foo` if its sources have changed
- if `_build/install/default/bin/foo.exe` hasn't previously been built, then you get an error.

Specifying `dune exec -- foo.exe` works consistently, but this has two problems:
- It's surprising on Windows (unless the user has done something really weird with the `PATHEXT` environment variable)
- It's not good for a cross-platform set-up since `dune exec -- foo.exe` will fail on a (Unix) system which installs `foo.exe` as `foo` with the executable bit set.

(Dune 2.4.0; Windows mingw32; OCaml 4.08.1)

**Comment by diml** [comment]
This is indeed a bug. In particular, `dune exec -- <prog>` should be consistent with `(run <prog>)` in dune files.

**Comment by Alizter** [comment]
On main (soon to be 3.21):
```
PS C:\Users\a\test_exe_3322> dune exec -- foo
Hello World

PS C:\Users\a\test_exe_3322> dune exec -- foo.exe
Error: Program 'foo.exe' not found!

PS C:\Users\a\test_exe_3322> dune exec -- bin\foo.exe
Hello World

PS C:\Users\a\test_exe_3322> dune exec -- .\_build\install\default\bin\foo.exe
Hello World
```

Here `bin/foo.ml` is a public executable.

The second behaviour is perhaps undesirable. Currently it errors regardless of whether it is built.

**Comment by dra27** [comment]
I think the second behaviour has been present for a little while, possibly even since Dune 3.x (I just double-checked that Dune 3.16 is indeed doing that in the switch for a deployment I have to hand):

```console
C:\Users\DRA\Documents\Signup\MOSS>dune exec -- payroll.exe --ping
Error: Program 'payroll.exe' not found!

C:\Users\DRA\Documents\Signup\MOSS>dune exec -- payroll --ping
pong
```

I think I opened this issue originally for this particular project because the second form didn't work and the first form did, and at some point it swapped round!

**Comment by Alizter** [comment]
Just for completeness:
```
PS C:\Users\a\test_exe_3322> dune clean
PS C:\Users\a\test_exe_3322> dune exec --cache disabled -- foo
Hello World

PS C:\Users\a\test_exe_3322> dune clean
PS C:\Users\a\test_exe_3322> dune exec --cache disabled -- foo.exe
Error: Program 'foo.exe' not found!

PS C:\Users\a\test_exe_3322> dune clean
PS C:\Users\a\test_exe_3322> dune exec --cache disabled -- .\bin\foo.exe
Hello World

PS C:\Users\a\test_exe_3322> dune clean
PS C:\Users\a\test_exe_3322> dune exec --cache disabled -- .\_build\install\default\bin\foo.exe
Hello World
```
So we are definitely trying to build every time and not just executing present artefacts.

I will mark this issue as a bug, since I think its sensible for the second form to work here.

**Comment by dra27** [comment]
I agree - the behaviour in 2020 (when I opened this) was a big papercut (needing to explicitly add `.exe` on Windows). The current behaviour is a small papercut.

Incidentally, for `(public_name foo)` I would expect `dune exec -- foo` to work on both on Unix and Windows (as it does in 3.x) and `dune exec -- foo.exe` only to work on Windows, consistent with how cmd/pwsh both work. I think it's consistent for it to be Windows-only behaviour because the installed `foo` binary is actually named `foo.exe` on Windows where the installed binary is just `foo` on Unix (i.e. it's different from the internal `.exe` suffix)
